proposals/p1964.md

```diff
@@ -0,0 +1,351 @@
+C++, Java, Swift to name a few. Whereas other languages that lack distinct
+character type, like Python use strings of length one to serve the same purpose
+a character data type. For more information see
+[Character Literals Wiki](https://en.wikipedia.org/wiki/Character_literal),
+[Character Literals DBpedia](https://dbpedia.org/page/Character_literal)
+
+## Proposal
+
+Put character literals in single quotes, like `'a'`. Character literals work
+like numeric literals:
+
+-   Every different literal value has its own type.
+-   The bit width is determined by the type of the variable the literal is
+    assigned to, not the literal itself. Follows the plan from #1934.
+
+We will not support:
+
+-   Multi-line literals
+-   "raw" literals (using #'x'#)
+-   Empty character literals (`''`)
+-   `\x` escape sequences
+-   Character literals that don't contain exactly one code point.
+-   ASCII Control codes (0...31), except when specified with an escape sequence.
+-   Whitespace characters other than word space: tab, line feed, carriage
+    return, form feed, and vertical tab. See
+    [String Literals: Escaping Sequence](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0199.md#escape-sequences-1)
+
+## Details
+
+-   A character literal is a sequence enclosed with single quotes delimiter ('),
+    of UTF-8 code units that may or may not be a valid encoding. This matches
+    [the UTF-8 encoding of Carbon source files](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0142.md#character-encoding).
+-   If a character literal encodes exactly one code point, then it supports
+    addition and subtraction. These operations produce another code point
+    literal, otherwise it will be rejected and produce an error.
+-   For most purposes a character should be viewed as representing a grapheme
+    cluster.
+-   Value-preserving implicit conversions from character literals to code point
+    or code unit types are permitted. In particular, a character literal
+    converts to a UTF-8 code unit if it is less than or equal to 0x7F, and
+    UTF-16 code unit if it is less than or equal to 0xFFFF.
+-   Support for `\u` escape sequences.
+-   Conversions from string literals to Unicode strings are implicit, even
+    though the numeric values of the encoding may change.
+
+Character literals cannot be written with these characters:
+
+-   New line
+-   Single quote (`'`), but can write `'` character as `'\''`
+-   Back-slash (`\`), except when used to form an escape sequence. See
+    [String Literals: Escaping Sequence](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0199.md#escape-sequences-1)
+-   Whitespace characters other than space (` `).
+-   ASCII control codes (0...31).
+
+### Types
+
+We will have the types `Char8`, `Char16`, and `Char32` representing code units
+in UTF-8, UTF-16, and UTF-32. `Char32` will also represent a valid code point.
```

Comment:
I wonder, should this should get clarified a bit to pin down that these don't support *all* code units in UTF-8 or UTF-16, but only those which map directly to the complete value of a code point?

Or should we keep these as types that can be formed from character literals but also might be used for representing utf-8 or utf-16 encoded strings? @zygoloid @josh11b